File: fetch_cdn.py
#!/usr/bin/env python3
"""Fetch ship default loadouts from Erkul.games CDN. Outputs ships_cdn.json."""
import json, time, urllib.request, urllib.error, zlib

CDN = "https://cdn.erkul.games"
HEADERS = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
    "Referer": "https://erkul.games/",
}

def fetch(url, retries=3):
    req = urllib.request.Request(url, headers=HEADERS)
    for attempt in range(retries):
        try:
            with urllib.request.urlopen(req, timeout=30) as r:
                return r.read()
        except urllib.error.HTTPError as e:
            if e.code in (429, 503) and attempt < retries - 1:
                time.sleep(2 ** attempt)
                continue
            raise
    raise RuntimeError(f"Failed after {retries} attempts: {url}")

def fetch_bin(url):
    return json.loads(zlib.decompress(fetch(url), -15))

def get_blobs():
    catalog = fetch_bin(f"{CDN}/LIVE/catalog.bin")
    group_path = next(g["indexPath"] for g in catalog["groups"] if g["kind"] == "ships")
    group = fetch_bin(f"{CDN}/LIVE/{group_path}")
    return group["blobs"], catalog["dataVersion"]

def extract_slots(ship_data):
    slots = []
    for slot in ship_data.get("slots", []):
        port = slot.get("portName", "")
        if "paint" in port.lower():
            continue
        item = slot.get("item") or {}
        if not item:
            continue
        name = (item.get("i18n") or {}).get("name") or item.get("className", "")
        if not name:
            continue
        slots.append({
            "portName": port,
            "type": item.get("type", ""),
            "category": item.get("category", ""),
            "itemName": name,
            "grade": item.get("grade", ""),
            "size": item.get("size", 0),
        })
    return slots

def main():
    print("Fetching ships index...")
    blobs, version = get_blobs()
    print(f"Game version: {version}  |  Ships: {len(blobs)}")

    ships = []
    for i, blob in enumerate(blobs, 1):
        url = f"{CDN}/LIVE/{blob['path']}"
        print(f"  [{i:3d}/{len(blobs)}] {blob['id']}", end=" ", flush=True)
        try:
            data = fetch_bin(url)
            name = (data.get("i18n") or {}).get("name") or blob["id"]
            slots = extract_slots(data)
            ships.append({"id": blob["id"], "name": name, "slots": slots})
            print(f"({len(slots)} slots)")
        except Exception as e:
            print(f"ERROR: {e}")
            ships.append({"id": blob["id"], "name": blob["id"], "slots": []})
        time.sleep(0.1)

    with open("ships_cdn.json", "w") as f:
        json.dump(ships, f)
    print(f"\nSaved {len(ships)} ships to ships_cdn.json")

if __name__ == "__main__":
    main()

File: fetch_cdn.js
#!/usr/bin/env node
const { inflateRawSync } = require("zlib");
const fs = require("fs");

const CDN = "https://cdn.erkul.games";
const HEADERS = {
  "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36",
  "Accept": "*/*",
  "Accept-Language": "en-US,en;q=0.9",
  "Referer": "https://erkul.games/",
  "Origin": "https://erkul.games",
  "sec-fetch-dest": "empty",
  "sec-fetch-mode": "cors",
  "sec-fetch-site": "same-site",
};

async function fetchBin(url, retries = 3) {
  for (let attempt = 0; attempt < retries; attempt++) {
    const resp = await fetch(url, { headers: HEADERS });
    if (resp.status === 429 || resp.status === 503) {
      await new Promise(r => setTimeout(r, 1000 * 2 ** attempt));
      continue;
    }
    if (!resp.ok) throw new Error(`HTTP ${resp.status} for ${url}`);
    const buf = await resp.arrayBuffer();
    return JSON.parse(inflateRawSync(Buffer.from(buf)).toString("utf-8"));
  }
  throw new Error(`Failed after ${retries} attempts: ${url}`);
}

async function getBlobs() {
  const catalog = await fetchBin(`${CDN}/LIVE/catalog.bin`);
  const groupPath = catalog.groups.find(g => g.kind === "ships").indexPath;
  const group = await fetchBin(`${CDN}/LIVE/${groupPath}`);
  return { blobs: group.blobs, version: catalog.dataVersion };
}

function extractSlots(shipData) {
  const slots = [];
  for (const slot of shipData.slots || []) {
    const port = slot.portName || "";
    if (port.toLowerCase().includes("paint")) continue;
    const item = slot.item;
    if (!item) continue;
    const name = item.i18n?.name || item.className || "";
    if (!name) continue;
    slots.push({
      portName: port,
      type: item.type || "",
      category: item.category || "",
      itemName: name,
      grade: item.grade || "",
      size: item.size || 0,
    });
  }
  return slots;
}

async function main() {
  console.log("Fetching ships index...");
  const { blobs, version } = await getBlobs();
  console.log(`Game version: ${version}  |  Ships: ${blobs.length}`);

  const ships = [];
  for (let i = 0; i < blobs.length; i++) {
    const blob = blobs[i];
    const url = `${CDN}/LIVE/${blob.path}`;
    process.stdout.write(`  [${String(i + 1).padStart(3)}/${blobs.length}] ${blob.id} `);
    try {
      const data = await fetchBin(url);
      const name = data.i18n?.name || blob.id;
      const slots = extractSlots(data);
      ships.push({ id: blob.id, name, slots });
      console.log(`(${slots.length} slots)`);
    } catch (e) {
      console.log(`ERROR: ${e.message}`);
      ships.push({ id: blob.id, name: blob.id, slots: [] });
    }
    await new Promise(r => setTimeout(r, 100));
  }

  fs.writeFileSync("ships_cdn.json", JSON.stringify({ version, ships }));
  console.log(`\nSaved ${ships.length} ships to ships_cdn.json`);
}

main().catch(e => { console.error(e); process.exit(1); });
